Compiling the mindcode project does not work since the release `"Prepared separate web application using Spring Web"`

Good morning @francois 

Thanks a lot for all your work!, 
I have tried to compile mindcode and used your _Native installation_ guide. Until to the commit from Apr 12, 2021 `Permit "indirect" sensor references` compiling works perfectly:

```
git checkout 03ecee928ce36ae33303e2b37f428c2f63634adb
mvn clean compile exec:java
… everything Ok
```
But if I move to the next release, [Github Commit: `Prepared separate web application using Spring Web`](https://github.com/francois/mindcode/commit/ab1ad052502e8f24a404cc064ca9840316b38534) I get a compiler error:
```
git checkout ab1ad052502e8f24a404cc064ca9840316b38534
mvn clean compile exec:java

[INFO] Reactor Summary:                                                                                                                                    
[INFO]                                                                                                                                                     
[INFO] mindcode-compiler 1.0.0 ............................ FAILURE [  3.806 s]                                                                            
[INFO] webapp 0.0.1-SNAPSHOT .............................. SKIPPED                                                                                        
[INFO] Mindcode 1.0.0-SNAPSHOT ............................ SKIPPED                                                                                        
[INFO] ------------------------------------------------------------------------                                                                            
[INFO] BUILD FAILURE                                                                                                                                       
[INFO] ------------------------------------------------------------------------                                                                            
[INFO] Total time:  4.383 s                                                                                                                                
[INFO] Finished at: 2021-04-20T10:43:38+02:00                                                                                                              
[INFO] ------------------------------------------------------------------------                                                                            
[ERROR] Failed to execute goal org.codehaus.mojo:exec-maven-plugin:3.0.0:java (default-cli) on project mindcode-compiler: The parameters 'mainClass' for go
al org.codehaus.mojo:exec-maven-plugin:3.0.0:java are missing or invalid -> [Help 1]
…
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginParameterException
```

Unfortunately, I don't know Maven at all, so it's a bit tricky to find the cause. About Java, I know the version is always important - it's fine, I guess:

```
java.exe -version
openjdk version "11.0.10" 2021-01-19
OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.10+9)
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 11.0.10+9, mixed mode)
C:\GitWork\Clones\GitHub\Game - Mindustry - mindcode\src [(ab1ad05...)]

mvn -version
Apache Maven 3.6.3 (cecedd343002696d0abb50b32b541b8a6ba2883f)
Maven home: C:\Portable\Java-Maven\apache-maven-3.6.3\bin\..
Java version: 11.0.10, vendor: AdoptOpenJDK, runtime: C:\Portable\Java-JDK\jdk-11
Default locale: de_CH, platform encoding: Cp1252
OS name: "windows 10", version: "10.0", arch: "amd64", family: "windows"
```

I can not attach a html-file with the output of `mvn clean compile exec:java -e` and `mvn clean compile exec:java -e -X`, therefore I have added a pdf, so that the colors are preserved - this is much easier to read.


[mvn clean compile exec_java -e -X.html.pdf](https://github.com/francois/mindcode/files/6342135/mvn.clean.compile.exec_java.-e.-X.html.pdf)
[mvn clean compile exec_java -e.html.pdf](https://github.com/francois/mindcode/files/6342136/mvn.clean.compile.exec_java.-e.html.pdf)

I really hope this is just a minor config issue. Please let me know if I can send you more information.

Thanks a lot, kind regards,
Thomas

### `mvn clean test` is bound to linux because of calling 'diff'

The tests use Linux `diff`, therefore I get this error on Windows:
```
[ERROR] info.teksol.mindcode.webapp.SamplesTest.validateSamples  Time elapsed: 0.042 s  <<< ERROR!
java.io.IOException: Cannot run program "diff": CreateProcess error=2, Das System kann die angegebene Datei nicht finden
```

I was able to fix it by replacing `diff` by `fc.exe`:

File: `\src\webapp\src\test\java\info\teksol\mindcode\webapp\SamplesTest.java`
```
final Process diff = new ProcessBuilder()
		.command("fc.exe", "-u", unoptimizedTarget.getAbsolutePath(), optimizedTarget.getAbsolutePath())
		.redirectOutput(diffTarget)
		.start();
```

I don't know if the Test-Engine is able to automatically distinguish between Linux and Windows, but Windows users can easily correct it with the above hint :-)

--- a/Dockerfile
+++ b/Dockerfile
@@ -6,8 +6,13 @@ RUN apt update && apt -y install maven
 
 COPY . .
 
-RUN mvn clean compile
+ENV SPRING_DATASOURCE_URL jdbc:postgresql://mindcode-db/mindcode_development
+ENV SPRING_DATASOURCE_USERNAME postgres
+ENV SPRING_DATASOURCE_PASSWORD pg_password
+
+# Skip tests because postgres is only available at runtime
+RUN ./mvnw install -Dmaven.test.skip
 
 EXPOSE 8080
 
-CMD mvn exec:java
+CMD java -classpath `find webapp -type f -name '*.jar' | tr '\n' ':'` info.teksol.mindcode.webapp.WebappApplication

--- a/README.markdown
+++ b/README.markdown
@@ -46,15 +46,18 @@ The Mindcode UI will now be running on localhost, port 8080. Visit http://localh
 
 1. Install Java 11+, Maven 3.6, and PostgreSQL
 2. Create a database in PostgreSQL named `mindcode_development`
-3. Set an environment variable with the PostgreSQL connection parameters:
+3. Set environment variables with the PostgreSQL connection parameters:
 ```sh
-export JDBC_DATABASE_URL='jdbc:postgresql://localhost/mindcode_development?user=POSTGRES_USER&password=POSTGRES_PASSWORD'
+export SPRING_DATASOURCE_URL=jdbc:postgresql://localhost/mindcode_development
+export SPRING_DATASOURCE_USERNAME=postgres_username
+export SPRING_DATASOURCE_PASSWORD=postgres_password
 ```
 
 Then run:
 
 ```sh
-mvn clean compile exec:java
+./mvnw install
+java -classpath $( find webapp -type f -name '*.jar' | tr '\n' ':' ) info.teksol.mindcode.webapp.WebappApplication
 ```
 
 The Mindcode UI will now be running on localhost, port 8080. Visit http://localhost:8080/ to interact with it.

--- a/docker-compose.yaml
+++ b/docker-compose.yaml
@@ -17,8 +17,6 @@ services:
   mindcode-web:
     build:
       context: .
-    environment:
-      JDBC_DATABASE_URL: jdbc:postgresql://mindcode-db/mindcode_development?user=postgres&password=pg_password
 
     depends_on:
       - mindcode-db

